### DISCUSSION TITLE
service un/bind commands are stateful. better to be idempotent ?

### DISCUSSION THREAD

**Discussion by andreas-kupries** [epinio/epinio#827] [General]
Un/binding services from/to apps is currently stateful.
I.e. binding a service multiple times to the same app generates errors after the first time (`already bound`).
Ditto unbinding a service multiple times (`not bound`).

Being in the guts of that to support un/bind for apps at rest I believe that
  - the code would be simpler if we were to go for idempotent. I.e. second un/bind simply does nothing.
  - might be simpler for users to handle as well (less nagging, simpler mental model)
  - should make use of epinio in scripts simpler also - script does not have to track un/bind state

It would however be an API change
  - no errors where currently expected
  -  fields like `WasBound` in API responses dropped (I know of this one, not sure about others)
  - etc.

Thoughts ?

(In a different community I am active in this is called `do nothing gracefully` (instead of an error)).

**Comment by manno** [comment]
We could return 304 Not Modified


I want to change how we handle status codes at some point. It seems complicated to  pass in a custom error handling funcs to support non-200 codes. And that’s in the client, so a framework probably would not help.

**Reply by andreas-kupries** [reply]
`304 not modified` sounds sensible. code complexity should still be less than what we have at the moment.

**Reply by manno** [reply]
Yes, however I just realized that's a redirect. (https://restfulapi.net/http-status-codes/)

I wanted to signal to the consumer, that the request is fine ok, but nothing changed.

POST is normally not idempotent, so we could just return 200, but then the consumer wouldn't know nothing happened.

So,  should the body contain the full resource json (200) or an empty body (304)?

**Reply by agracey** [reply]
I think it should return a 200 as if it had made the change. That'll be less complexity for external tooling. Which 304 is likely more "correct", the call did succeed, it just didn't change the internal state so 200 is fine IMO.

**Reply by andreas-kupries** [reply]
Maybe https://restfulapi.net/http-status-204-no-content/ then ?
Seems to be related to 304. That said, I am perfectly fine with 200.

**Comment by andreas-kupries** [comment]
The general trend in responses seems to be that idempotent is ok/better (@jimmykarily in the slack), with main discussion on the exact status code to return.

I'll leave the discussion open until tomorrow, in case of late comers. If there is no change tomorrow I will add idempotency to my work planning document.

**Reply by manno** [reply]
I lean more towards 200 now, but we should make sure the resource status includes a created/updated timestamp (maybe from a related k8s resource).

**Comment by andreas-kupries** [comment]
done, and going for idempotent.
